### PR TITLE
[SPARK-30379][Core] Avoid OOM when using collection accumulator

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch proposes to only convert first few elements of collection accumulators in `LiveEntityHelpers.newAccumulatorInfos`.

### Why are the changes needed?

One Spark job on our cluster uses collection accumulator to collect something and has encountered an exception like:

```
java.lang.OutOfMemoryError: Java heap space
    at java.util.Arrays.copyOf(Arrays.java:3332)
    at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:124)
    at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:448)
    at java.lang.StringBuilder.append(StringBuilder.java:136)
    at java.lang.StringBuilder.append(StringBuilder.java:131)
    at java.util.AbstractCollection.toString(AbstractCollection.java:462)
    at java.util.Collections$UnmodifiableCollection.toString(Collections.java:1035)
    at org.apache.spark.status.LiveEntityHelpers$$anonfun$newAccumulatorInfos$2$$anonfun$apply$3.apply(LiveEntity.scala:596)
    at org.apache.spark.status.LiveEntityHelpers$$anonfun$newAccumulatorInfos$2$$anonfun$apply$3.apply(LiveEntity.scala:596)
    at scala.Option.map(Option.scala:146)
    at org.apache.spark.status.LiveEntityHelpers$$anonfun$newAccumulatorInfos$2.apply(LiveEntity.scala:596)
    at org.apache.spark.status.LiveEntityHelpers$$anonfun$newAccumulatorInfos$2.apply(LiveEntity.scala:591)
```

`LiveEntityHelpers.newAccumulatorInfos` converts `AccumulableInfo`s to `v1.AccumulableInfo` by calling `toString` on accumulator's value. For collection accumulator, it might take much more memory when in string representation, for example, collection accumulator of long values, and cause OOM (in this job, the driver memory is 6g).

Looks like the results of `newAccumulatorInfos` are used in api and ui. For such usage, it also does not make sense to have very long string of complete collection accumulators.

### Does this PR introduce any user-facing change?

Yes. Collection accumulator now only shows first few elements in api and ui.

### How was this patch tested?

Unit test.

Manual test. Launched a Spark shell, ran:
```scala
val accum = sc.collectionAccumulator[Long]("Collection Accumulator Example")
sc.range(0, 10000, 1, 1).foreach(x => accum.add(x))
accum.value
```

Screen Shot 2019-12-30 at 2 03 43 PM

### CODE REVIEW

**Comment by Ngone51** [comment]
retest this please.

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala [hunk 1]
```diff
@@ -625,10 +626,19 @@ private class SchedulerPool(name: String) extends LiveEntity {
...
 }
 
-private object LiveEntityHelpers {
+private[spark] object LiveEntityHelpers {
 
   private val stringInterner = Interners.newWeakInterner[String]()
 
+  private def accuValuetoString(value: Any): String = value match {
+    case list: java.util.List[_] =>
+      if (list.size() > 5) {
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: Would it be better if we add a config for this?

Comment (viirya) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: I'm not sure if a config here is useful. Will we want to set more or less elements of collection accumulators in api or ui? cc @cloud-fan

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: Why don't we just use existing configuration `spark.sql.debug.maxToStringFields`?

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: Oops, it's not only SQL.

Comment (viirya) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: yea. :)

**Review by cloud-fan — APPROVED** [review]
makes sense to me

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala [hunk 2]
```diff
@@ -625,10 +626,19 @@ private class SchedulerPool(name: String) extends LiveEntity {
 
 }
 
-private object LiveEntityHelpers {
+private[spark] object LiveEntityHelpers {
 
   private val stringInterner = Interners.newWeakInterner[String]()
 
+  private def accuValuetoString(value: Any): String = value match {
+    case list: java.util.List[_] =>
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: Can we add a JIRA ID and why we only abbreviate this case alone?

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala [hunk 3]
```diff
@@ -625,10 +626,19 @@ private class SchedulerPool(name: String) extends LiveEntity {
...
+      if (list.size() > 5) {
+        "[" + list.asScala.take(5).mkString(", ") + "...]"
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: Can we make the output look like `truncatedString` just to make it look consistent?

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala (hunk 1 above)
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: Does 5 looks good enough in UI? I haven't tested it by myself but thought it's faster just to ask.

**Review by HyukjinKwon — APPROVED** [review]
Looks okay.

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala (hunk 1 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: This is a number for avoiding OOM issue. For UI, good or not might be case by case. Although it sounds okay to add a config, I feel it is not a big deal to adjust this number in UI and api. I suggest to leave it as 5 and if we need to configure it, we can add a config later.

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala (hunk 2 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: Yes. Good advice.

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala (hunk 3 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: Yes, it is better to let users know how many more items there are.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
Thanks for review. Will merge this later today.

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala [hunk 4]
```diff
@@ -641,8 +654,8 @@ private object LiveEntityHelpers {
         new v1.AccumulableInfo(
           acc.id,
           acc.name.map(weakIntern).orNull,
-          acc.update.map(_.toString()),
-          acc.value.map(_.toString()).orNull)
+          acc.update.map(accuValuetoString),
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: Just found that in UI, the tasks of Stage tab shows `update` too. So we also need to truncate update value.

Comment (viirya) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: Shown in the pr description.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**